rewrite-core/src/main/java/org/openrewrite/Metadata.java

```diff
@@ -15,13 +15,94 @@
  */
 package org.openrewrite;
 
-import com.fasterxml.jackson.annotation.JsonIdentityInfo;
-import com.fasterxml.jackson.annotation.JsonProperty;
-import com.fasterxml.jackson.annotation.JsonTypeInfo;
-import com.fasterxml.jackson.annotation.ObjectIdGenerators;
+import com.fasterxml.jackson.annotation.*;
+import com.fasterxml.jackson.annotation.JsonAutoDetect.Visibility;
+import org.openrewrite.internal.lang.Nullable;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+import static java.util.Collections.emptyList;
 
 @Incubating(since = "2.0.0")
-@JsonIdentityInfo(generator = ObjectIdGenerators.IntSequenceGenerator.class, property = "@ref")
-@JsonTypeInfo(use = JsonTypeInfo.Id.CLASS, property = "@c")
-public interface Metadata {
+@JsonAutoDetect(
+        fieldVisibility = Visibility.ANY,
+        getterVisibility = Visibility.NONE,
+        setterVisibility = Visibility.NONE
+)
+public class Metadata {
```

Comment:
Look at this closely during review. The interface to extend is now `Metadata.Datum`. and `Metadata` is a holder of this collection. This way we can develop functionality around capturing and updating metadata without using a series of `default` methods on the `Tree` interface and further polluting the API namespace of `Tree`.

Reply:
The name may not be great, since Kotlin has a `kotlin.Metadata` that is implicitly imported. The IDE experience for using `org.openrewrite.Metadata` is a bit weird. You write out the fully qualified name on first use and then use the completion action to add the import. Otherwise IntelliJ assumes you want to use the Kotlin `Metadata` type.

Reply:
I think because there is a potential conflict when using generalized names like MetaData or Context, how about if we prefix the name? `TreeMetaData` or `NodeMetaData`?